docs/coding-guidelines.md

```diff
@@ -299,3 +300,42 @@ stack potentially unboundedly. Instead use iteration e.g. via `loop` or `for`.
 - https://en.wikipedia.org/wiki/Tail_call
 - https://stackoverflow.com/questions/65948553/why-is-recursion-not-suggested-in-rust
 - https://stackoverflow.com/questions/59257543/when-is-tail-recursion-guaranteed-in-rust
+
+## Allow Correlating Asynchronous Responses to Their Requests
+
+In an asynchronous context, it is important to enable users to determine the correlation between a
+response and a previous request. For example, if a user requests two new connections to the same
+peer, they should be able to match each new connection to the corresponding previous connection
+request without having to guess.
+
+When providing a **method** where the response to that method is delivered asynchronously through an
+event, either synchronously return a request ID which is later on contained in the asynchronous
+response event, or synchronously return a `Future` that eventually resolves into the response.
+
+``` rust
+fn my_method() -> Id {
+  // ...
+}
+```
```

Comment:
Would it make sense to remove this and be consistent with commands where the caller always specifies the ID? Only `Transport::dial` currently behaves like stated here and we already said we wanted to fix this! :)

I'd be in favor of removing this from the guidelines (or rather, say that we want to accept the ID) and opening an issue to fix `Transport::dial`.

Reply:
> with commands where the caller always specifies the ID?

I have to give this more thought. In other words, I am not sure this is the way to go in all situations.

Do you feel strongly about removing this @thomaseizinger? I would prefer proceeding here. I see this pull request as a net-positive at this point.

Reply:
I think it makes sense for anything that is public API in our library. Internally, where we control all usages of an ID, it doesn't matter as much.

The reason for that is that the majority of our public API is command-based. Anything command-based needs to create the ID ahead of time. Any component underneath that creates its own IDs causes friction here because it doesn't compose well with commands.

Reply:
I am a bit torn. There is certainly value in documenting this but if we document something, I'd prefer if we can back the written words 100%, e.g. have strong convictions that this is the way to go. Everything else is just confusing to the reader.

Reply:
Would removing this section be a middle ground? i.e. only document that commands should come with a user-generated ID and not talk about functions at all?

Reply:
Done with 12e61b63.